### PR TITLE
Fix some nullable annotations / null handling in System.Text.Json

### PR DESCRIPTION
**Pull request by stephentoub**
The `JsonConverter<T>`s for reference type `T`s need to be able to deal with null. Fix the nullable annotations accordingly. This then led to highlghting that the implementation can null ref if a dictionary returns a null key, so adding appropriate validation of that, too.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValue.CreateOverloads.cs [hunk 1]
```diff
@@ -225,7 +225,7 @@ public partial class JsonValue
         /// <param name="options">Options to control the behavior.</param>
         /// <returns>The new instance of the <see cref="JsonValue"/> class that contains the specified value.</returns>
         [return: NotNullIfNotNull(nameof(value))]
-        public static JsonValue? Create(string? value, JsonNodeOptions? options = null) => value != null ? new JsonValueTrimmable<string>(value, JsonMetadataServices.StringConverter) : null;
+        public static JsonValue? Create(string? value, JsonNodeOptions? options = null) => value != null ? new JsonValueTrimmable<string?>(value, JsonMetadataServices.StringConverter) : null;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValue.CreateOverloads.cs:228]: Not sure I understand why this change was made.

**Review by eiriktsarpalis — APPROVED** [review]
Thanks 👍

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValue.CreateOverloads.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValue.CreateOverloads.cs:228]: Previously it was ok with `JsonValueTrimmable<string>` because even though `value` is a `string?`, this ctor is only used after `value != null`.  But now `StringConverter` uses type `string?` rather than `string`, so the `T` needs to be `string?`.

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs [hunk 2]
```diff
@@ -913,7 +913,7 @@ public abstract partial class JsonConverter<T> : System.Text.Json.Serialization.
             T value,
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:913]: Seems worth following up on to see whether we can make it work reasonably.

[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:913]: @eerhardt is querying for NRT attributes using reflection something we can assume to be reliable in the context of the trimmer and NativeAOT?

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:913]: There is a [feature switch](https://github.com/dotnet/runtime/blob/main/docs/workflow/trimming/feature-switches.md) that will trim away the nullability attributes: NullabilityInfoContextSupport. When this switch is set to `false`, querying using NullabilityInfoContext will throw an exception saying that it isn't supported.

The switch is only defaulted in places that are super size sensitive, like Blazor WASM and iOS/Android apps.

See

https://github.com/dotnet/runtime/blob/b05b4e44c00f7d8e32aa6b839b842f7ddd7d503b/src/libraries/System.Private.CoreLib/src/System/Reflection/NullabilityInfoContext.cs#L262-L268

[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:913]: Thanks, by the sound if it we probably shouldn't be determining runtime behavior based on the presence of that attribute.
